### ISSUE TITLE
`decompressFromBase64` can return `null`

### ISSUE THREAD

**Issue by emewjin** [pieroxy/lz-string#184]
version: 1.5.0

[codesandbox](https://codesandbox.io/p/sandbox/inspiring-merkle-wn8kqt?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clq68fm3o00063b84j6v7jj7a%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clq68fm3o00023b844yhk8s4k%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clq68fm3o00033b84snwklf98%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clq68fm3o00053b84r5um1g5q%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clq68fm3o00023b844yhk8s4k%2522%253A%257B%2522id%2522%253A%2522clq68fm3o00023b844yhk8s4k%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clq68fm3o00053b84r5um1g5q%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clq68fm3o00043b84l48qs41l%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clq68fm3o00053b84r5um1g5q%2522%252C%2522activeTabId%2522%253A%2522clq68fm3o00043b84l48qs41l%2522%257D%252C%2522clq68fm3o00033b84snwklf98%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clq68fm3o00033b84snwklf98%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) 

`decompressFromBase64` can return `null` but it's type definition is `export function decompressFromBase64(input: string): string;`.
![image](https://github.com/pieroxy/lz-string/assets/76927618/52e3f5b2-b44e-4afb-926d-b39ef23244c4)

It should be `export function decompressFromBase64(input: string): string | null;`

```js
console.log(decompressFromBase64('')) // null
```

**Comment by HelloLudger** [comment]
I never understood why converting an empty string would return `null` instead of `""`. 

My emotional expectation would be that `""` is the return of an empty input and `null` is returned in the event of an error.

**Comment by wkrick** [comment]
@HelloLudger - I think the reason is because an empty string isn't a valid Base64 encoded string.  There's no inputs into the Base64 encoding algorithm that generate an empty string so trying to decode an empty string makes no sense.  So returning null from the function is a reasonable way to indicate an error due to invalid input.

**Comment by HelloLudger** [comment]
@wkrick that would be an explaination, but I found no arguments to support that `""` is no valid Base64 (or the other way around). 

- Stack Overflow: https://stackoverflow.com/questions/15212127/is-an-empty-string-valid-base64-encoded-data-of-zero-bytes-length
- `window.atob` and `window.btoa` (who named these functions???) also convert an empty string input to an empty string output.

**Comment by wkrick** [comment]
@HelloLudger 

Interesting.  The test cases proposed in the RFC clearly indicate that encoding an empty string results in an empty string...

`BASE64("") = ""`

I stand corrected.

**Comment by Rycochet** [comment]
Two issues here, first is code output (and a breaking change) - an empty string is valid for an empty result, an error should be different. Second is that the typings don't match what they should be and isn't breaking (and is also not going to change if / when the code change)

For the empty string result we do need to check it's not breaking in the other implementations otherwise it can't be changed for legacy reasons (and would need to be part of any future plans etc)

**Comment by Rycochet** [comment]
Just realised that the typings are correct for v2 - and had a look at some of the implementations for other languages, which have reproduced this behaviour faithfully, so it looks like a very breaking change!

**Comment by HelloLudger** [comment]
Yes... nothing that could be done without breaking. Only idea I have (and read here before in another context) would be adding mirrored functions with a new systematic of return values. (E.g. `compressToBase64Preview`)
